### PR TITLE
Issue #493 - consider comparators for Quantity data types

### PR DESCRIPTION
**Pull request by michaelwschroeder**
Signed-off-by: Mike Schroeder <mschroed@us.ibm.com>

### CODE REVIEW

**Comment by michaelwschroeder** [comment]
See the design doc attached to issue #493 for detailed description of changes.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 1]
```diff
@@ -128,9 +132,10 @@ public void testSearchQuantity_Quantity_withPrefix_GT() throws Exception {
 
     @Test
     public void testSearchQuantity_Quantity_withPrefix_LE() throws Exception {
+        // Implicit range of indexed 'Quantity' value is 24.5 to 25.5
         assertSearchDoesntReturnSavedResource("Quantity", "le24|http://unitsofmeasure.org|s");
         assertSearchDoesntReturnSavedResource("Quantity", "le24.4999||s");
-        assertSearchReturnsSavedResource("Quantity", "le24.5||s");
+        assertSearchDoesntReturnSavedResource("Quantity", "le24.5||s");
         assertSearchReturnsSavedResource("Quantity", "le25||s");
```
Comment (tbieste) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java:139]: Would be good to have both edges of where the resource is returned vs not returned. ("le 24.9999||s"?)

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java [hunk 2]
```diff
@@ -139,42 +144,40 @@ public void testSearchQuantity_Quantity_withPrefix_LE() throws Exception {
...
+        // Implicit range of indexed 'Quantity' value is 24.5 to 25.5
         assertSearchReturnsSavedResource("Quantity", "ge24|http://unitsofmeasure.org|s");
         assertSearchReturnsSavedResource("Quantity", "ge24.4999||s");
         assertSearchReturnsSavedResource("Quantity", "ge24.5||s");
         assertSearchReturnsSavedResource("Quantity", "ge25||s");
         assertSearchReturnsSavedResource("Quantity", "ge25.4999||s");
-        // We need to track whether the high is inclusive or not to get this right
-        // For decimals, the high is exclusive, but for Range the high is inclusive
-//        assertSearchDoesntReturnSavedResource("Quantity", "ge25.5||s");
+        assertSearchDoesntReturnSavedResource("Quantity", "ge25.5||s");
```
Comment (tbieste) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java:153]: Would be good to have both edges of where the resource is returned vs not returned. ("le 25.0001||s"?)

**Review by tbieste — APPROVED** [review]

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 1 above)
Comment (michaelwschroeder) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java:139]: In this case, the edges are search values of 24.5 (won't return resource) and a value greater than 24.5 (will return a resource). Already have the 24.5 case - added a case of 24.5001.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java (hunk 2 above)
Comment (michaelwschroeder) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchQuantityTest.java:153]: In this case, the edges are search values of 25.5 (won't return resource) and a value less than 25.5 (will return a resource). Already have a 25.5 case and a 25.4999 case, so I think the edges are covered.

**Comment by tbieste** [comment]
LGTM

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java [hunk 3]
```diff
@@ -506,6 +507,8 @@ public boolean visit(java.lang.String elementName, int elementIndex, Money money
             p.setResourceType(resourceType);
             p.setName(searchParamCode);
             p.setValueNumber(money.getValue().getValue());
+            p.setValueNumberLow(NumberParmBehaviorUtil.generateLowerBound(money.getValue().getValue()));
+            p.setValueNumberHigh(NumberParmBehaviorUtil.generateUpperBound(money.getValue().getValue()));
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java:511]: should this be the NewNumberParmBehaviorUtil ? does that have the lower/upper bound stuff?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java [hunk 4]
```diff
@@ -551,8 +554,29 @@ public boolean visit(java.lang.String elementName, int elementIndex, Quantity qu
...
+                    int scale = value.scale() < 0 ? 9 : value.scale()+9;
+                    valueLow = value.add(new BigDecimal("1e-" + scale));
+                } else if (QuantityComparator.LESS_OR_EQUALS.equals(quantity.getComparator())) {
+                    // Range high value is quantity value
+                    valueHigh = value;
+                } else if (QuantityComparator.LESS_THAN.equals(quantity.getComparator())) {
+                    // Range high value is quantity value minus 1e-<scale+9>
+                    int scale = value.scale() < 0 ? 9 : value.scale()+9;
+                    valueHigh = value.subtract(new BigDecimal("1e-" + scale));
+                }
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java:575]: Isn't the Else If actually an Else? (e.g. it wasn't the previous 3 so it must be an else.  It'd skip a comparison.

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/NewNumberParmBehaviorUtilTest.java [hunk 5]
```diff
@@ -133,44 +133,48 @@ public void testPrecisionWithExact() throws FHIRPersistenceException {
...
+                + "OR Basic.NUMBER_VALUE IS NOT NULL AND Basic.NUMBER_VALUE_LOW IS NULL)";
         runTest(queryParm, expectedBindVariables, expectedSql);
 
         // ge - Greater than Equal
         queryParm             = generateParameter(SearchConstants.Prefix.GE, null, "1e3");
         expectedBindVariables = new ArrayList<>();
         expectedBindVariables.add(new BigDecimal("1E+3"));
-        expectedBindVariables.add(new BigDecimal("1E+3"));
-        expectedSql = "(Basic.NUMBER_VALUE >= ? OR Basic.NUMBER_VALUE_HIGH >= ?)";
+        expectedBindVariables.add(new BigDecimal("5E+2"));
```
Comment (prb112) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/NewNumberParmBehaviorUtilTest.java:153]: How does this end up with 5E+2?

Comment (prb112) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/NewNumberParmBehaviorUtilTest.java:153]: OK - got it - nevermind (it's the scale) and  yes this works.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/NewNumberParmBehaviorUtil.java [hunk 6]
```diff
@@ -236,15 +251,41 @@ public static void buildEqualsRangeClause(WhereFragment whereClauseSegment, Stri
...
-        whereClauseSegment.col(tableAlias, columnBase).gte().bind(lowerBound.subtract(factor)); // -10% of the Lower Bound
+        
+        whereClauseSegment.col(tableAlias, columnBase + _HIGH).gte().bind(approximateLowerBound);
         whereClauseSegment.and();
-        whereClauseSegment.col(tableAlias, columnBase).lt().bind(upperBound.add(factor)); // +10% of the Upper Bound
+        whereClauseSegment.col(tableAlias, columnBase + _LOW).lte().bind(approximateUpperBound);
+        
         whereClauseSegment.or();
-        whereClauseSegment.col(tableAlias, columnBase + _LOW).lte().bind(upperBound);
+        whereClauseSegment.col(tableAlias, columnBase).isNull();
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/NewNumberParmBehaviorUtil.java:264]: I've stared at this SQL for a while, it feels decently complex, could you inline some descriptions of the conditions being checked.  E.g. an example inline.  

It think it's OK, it'd just be great to not recreate the knowledge next time we're in the code.

**Review by prb112 — APPROVED** [review]
LGTM - left one javadoc request.  Also the failing IT should be resolved. (may just need a rebuild).

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java (hunk 3 above)
Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java]: Good catch - changed references to NewNumberParmBehaviorUtil (methods are same as old class)

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java (hunk 4 above)
Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java:575]: changed to just Else

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/NewNumberParmBehaviorUtilTest.java [hunk 7]
```diff
@@ -281,11 +277,18 @@ public void testPrecisionWithApprox() throws FHIRPersistenceException {
...
+        expectedBindVariables.add(new BigDecimal("110.5")); // V <= upperBound + factor
+        expectedBindVariables.add(new BigDecimal("89.5"));  // V >= lowerBound - factor
+        expectedBindVariables.add(new BigDecimal("110.5")); // V <= upperBound + factor
+        expectedBindVariables.add(new BigDecimal("89.5"));  // V >= lowerBound - factor
+        expectedBindVariables.add(new BigDecimal("110.5")); // V <= upperBound + factor
+        String expectedSql = "(Basic.NUMBER_VALUE_HIGH >= ? AND Basic.NUMBER_VALUE_LOW <= ? "
+                + "OR Basic.NUMBER_VALUE IS NULL AND (Basic.NUMBER_VALUE_HIGH >= ? AND Basic.NUMBER_VALUE_HIGH <= ? "
+                + "OR Basic.NUMBER_VALUE_LOW >= ? AND Basic.NUMBER_VALUE_LOW <= ?) "
+                + "OR Basic.NUMBER_VALUE IS NOT NULL AND (Basic.NUMBER_VALUE_HIGH >= ? AND Basic.NUMBER_VALUE_LOW IS NULL "
+                + "OR Basic.NUMBER_VALUE_LOW <= ? AND Basic.NUMBER_VALUE_HIGH IS NULL))";
```
[the remark this answers is not in the document]
Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/util/NewNumberParmBehaviorUtilTest.java:291]: added comments

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/NewNumberParmBehaviorUtil.java (hunk 6 above)
Comment (michaelwschroeder) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/type/NewNumberParmBehaviorUtil.java:277]: added comments with descriptions and example

**Pull request merged**